### PR TITLE
Don't save data on immutable files to per-project disk cache

### PR DESCRIPTION
**Pull request by ladipro**
Contributes to #8635

### Context

With #8688 we no longer need to cache data on immutable framework files on disk as obtaining this information is cheap and does not require I/O.

### Changes Made

Made `SystemState` (the RAR cache) not add such files to the per-instance dictionary. This dictionary is serialized to disk as `<project-file>.AssemblyReference.cache` so the change effectively makes the cache file smaller, saving on serialization and deserialization time.

Also refactored `DeserializeCache` into a generic method instead of taking a `Type` parameter.

For a simple ASP.NET Core app with one project reference and one package reference, the size of the cache file is reduced from 142 kB to 2 kB. Projects with no references other than the SDK will not have the cache file created at all.

### Testing

Existing unit tests, manual measurements.

### Notes

The actual "Don't read the file if the data is already in memory" change will be in a separate upcoming PR.

### CODE REVIEW

File: src/Tasks/SystemState.cs [hunk 1]
```diff
@@ -210,6 +210,16 @@ internal FrameworkName FrameworkNameAttribute
...
+
+            /// <summary>
+            /// The last-modified value to use for immutable framework files which we don't do I/O on.
+            /// </summary>
+            internal static DateTime ImmutableFileLastModifiedMarker => DateTime.MaxValue;
+
+            /// <summary>
+            /// It is wasteful to persist entries for immutable framework files.
+            /// </summary>
+            internal bool IsWorthPersisting => lastModified != ImmutableFileLastModifiedMarker;
```
Comment (yuehuang010) [inline src/Tasks/SystemState.cs:222]: If the file doesn't exists, would it be min or max datetime?
Likewise, if the file isn't an assembly?

**Review by JanKrivanek — APPROVED** [review]
Looks good!

File: src/Tasks/SystemState.cs (hunk 1 above)
Comment (ladipro) [reply src/Tasks/SystemState.cs:222]: This is a positive cache of assemblies, I don't think we create an entry for a non-existent file or a file that is not an assembly. In general we use `DateTime.MinValue` to represent "missing", though.

**Review by rainersigwald — APPROVED** [review]

**Pull request merged**